x-pack/plugins/security_solution/server/lib/entity_analytics/risk_score/risk_score_synchronous_upgrader.ts

```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { Logger } from '@kbn/logging';
+
+export class RiskScoreSynchronousUpgrader {
+  private constructor() {} // private constructor to prevent instantiation
+  private static upgradesConducted: Record<string, Promise<void>> = {};
+
+  /**
+   * An async function that ensures that {@link upgradeFunction}s are executed synchronously,
+   * no matter how many times they are called.
+   *
+   * @returns A promise returned by the {@link upgradeFunction} of a previous invocation of
+   * this {@link upgrade} function, or the promise returned by the {@link upgradeFunction} if one hasn't been provided yet.
+   */
+  static async upgrade(namespace: string, logger: Logger, upgradeFunction: () => Promise<void>) {
+    if (!RiskScoreSynchronousUpgrader.upgradesConducted[namespace]) {
+      RiskScoreSynchronousUpgrader.upgradesConducted[namespace] = upgradeFunction().catch((err) => {
```

Comment:
Are we using this value anywhere? If not, we could probably simplify to just setting a boolean here, at least until we need something more sophisticated.

Reply:
By "this value", are you referring to the `RiskScoreSynchronousUpgrader.upgradesConducted[namespace] = upgradeFunction()...` portion?

If so, yes, we are utilizing that value on line 29. That return always runs, while lines 22-28 simultaneously invoke the promise and set that promise as a value within `upgradesConducted`.